server/api/image.py

```diff
@@ -94,6 +98,84 @@ def find(self, params):
+        File = self.model('file')
+        Image = self.model('image', 'isic_archive')
+
+        if 'filter' in params:
+            query = querylang.astToMongo(json.loads(params['filter']))
+        else:
+            query = {}
+            if 'datasetId' in params:
+                try:
+                    query.update({'folderId': ObjectId(params['datasetId'])})
+                except InvalidId:
+                    raise RestException(
+                        'Invalid "folderId" ObjectId: %s' % params['datasetId'])
+
+        user = self.getCurrentUser()
+        downloadFileName = 'ISIC-images'
+
+        def stream():
+            datasetCache = dict()
+            zipGenerator = ziputil.ZipGenerator(downloadFileName)
+
+            for image in Image.filterResultsByPermission(
+                    # TODO: exclude additional fields from the cursor
+                    Image.find(query, sort=[('name', 1)], fields={'meta': 0}),
+                    user=user, level=AccessType.READ, limit=0, offset=0):
+                datasetId = image['folderId']
+                if datasetId not in datasetCache:
+                    datasetCache[datasetId] = Dataset.load(
+                        datasetId, force=True, exc=True)
+                dataset = datasetCache[datasetId]
+                imageFile = Image.originalFile(image)
+                imageFileGenerator = File.download(imageFile, headers=False)
+                for data in zipGenerator.addFile(
+                        imageFileGenerator,
+                        path=os.path.join(dataset['name'], imageFile['name'])):
+                    yield data
+
+            for dataset in six.viewvalues(datasetCache):
+                licenseText = mail_utils.renderTemplate(
+                    'license_%s.mako' % dataset['meta']['license'])
+                attributionText = mail_utils.renderTemplate(
+                    'attribution_%s.mako' % dataset['meta']['license'],
+                    {
+                        'work': dataset['name'],
+                        'author':
+                            dataset['meta']['attribution']
+                            if not dataset['meta']['anonymous']
+                            else 'Anonymous'
+                    })
+                for data in zipGenerator.addFile(
+                        lambda: [licenseText],
+                        path=os.path.join(dataset['name'], 'license.txt')):
+                    yield data
+                for data in zipGenerator.addFile(
+                        lambda: [attributionText],
+                        path=os.path.join(dataset['name'], 'attribution.txt')):
+                    yield data
+
+            yield zipGenerator.footer()
+
+        setResponseHeader('Content-Type', 'application/zip')
+        setResponseHeader(
+            'Content-Disposition',
+            'attachment; filename="%s.zip"' % downloadFileName)
+        return stream
```

Comment:
IMO, this is just a corner case, not an error. As long as it doesn't crash, I'd say it's fine to make an empty ZIP, to keep the return type in all non-error cases the same (`application/zip`).